src/plone/app/multilingual/browser/translate.py

```diff
@@ -1,53 +1,53 @@
 from Acquisition import aq_inner
 from plone.app.multilingual import _
-from plone.app.multilingual.interfaces import IMultiLanguageExtraOptionsSchema
+from plone.app.multilingual.interfaces import IExternalTranslationService
 from plone.app.multilingual.interfaces import ITranslationManager
 from plone.app.uuid.utils import uuidToObject
 from plone.base.interfaces import ILanguage
-from plone.registry.interfaces import IRegistry
 from plone.uuid.interfaces import IUUID
 from Products.Five import BrowserView
+from zope.component import getUtilitiesFor
 from zope.component import getUtility
 
 import json
-import urllib
-
-
-def google_translate(question, key, lang_target, lang_source):
-    length = len(question)
-    translated = ""
-    url = "https://www.googleapis.com/language/translate/v2"
-    temp_question = question
-    while length > 400:
-        temp_question = question[:399]
-        index = temp_question.rfind(" ")
-        temp_question = temp_question[:index]
-        question = question[index:]
-        length = len(question)
-        data = {
-            "key": key,
-            "target": lang_target,
-            "source": lang_source,
-            "q": temp_question,
-        }
-        params = urllib.parse.urlencode(data)
-
-        retorn = urllib.request.urlopen(url + "?" + params)
-        translated += json.loads(retorn.read())["data"]["translations"][0][
-            "translatedText"
+
+
+def translate_text(original_text, source_language, target_language, service=None):
+    """translate the text"""
+
+    if service is not None:
+        # if an specific adapter is requested, use it if available
+
+        adapter = getUtility(IExternalTranslationService, name=service)
+        if not adapter.is_available():
+            return None
+
+        adapters = [adapter]
+
+    else:
+
+        adapters = [
+            adapter
+            for _, adapter in getUtilitiesFor(IExternalTranslationService)
+            if adapter.is_available()
         ]
 
-    data = {
-        "key": key,
-        "target": lang_target,
-        "source": lang_source,
-        "q": temp_question,
-    }
-    params = urllib.parse.urlencode(data)
+    sorted_adapters = sorted(adapters, key=lambda x: x.order)
 
-    retorn = urllib.request.urlopen(url + "?" + params)
-    translated += json.loads(retorn.read())["data"]["translations"][0]["translatedText"]
-    return json.dumps({"data": translated})
+    for adapter in sorted_adapters:
+        available_languages = adapter.available_languages()
+        if (
+            not available_languages
+            or (source_language, target_language) in available_languages
+        ):
+            translation = adapter.translate_content(
```

Comment:
IMHO it is better to handle this loop using a try/except pattern, so that the whole service is not repeated if you try to translate something like empty text.